File: kotlin-node/karakum/injections/injectAgentOptionsPort.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node, context) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("https.d.ts")) return []

    if (
        ts.isInterfaceDeclaration(node)
        && node.name.text === "AgentOptions"
        && context.type === karakum.InjectionType.MEMBER
    ) {
        return ["override var port: dynamic"]
    }

    return []
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/injections/injectAgentOptionsPort.kt
package node.karakum.injections

import io.github.sgrishchenko.karakum.extension.InjectionType
import io.github.sgrishchenko.karakum.extension.createInjection
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import arrow.core.raise.Raise
import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.MEMBER
import typescript.isInterfaceDeclaration

val injectAgentOptionsPort = createInjection { node, context, _ ->
    nullable {
        ensure(context.type == InjectionType.MEMBER)

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("https.d.ts"))

        ensure(isInterfaceDeclaration(node))
        ensure(node.name.text == "AgentOptions")

        arrayOf(
            """
                @Suppress("VAR_TYPE_MISMATCH_ON_OVERRIDE")
                override var port: Double
            """.trimIndent()
        )
    }
}
